### PR TITLE
Fix makefiles' zfs-functions

### PR DESCRIPTION
**Pull request by rlaager**
### Motivation and Context
zfs-functions now has substitutions like `@bindir@`. If this file is not rebuilt after `./configure`, then it has the wrong paths, which leads to serious breakage. See #9443

### Description
The underlying issue here is that zfs-functions ends up in the tarball generated by `make dist`. This shouldn't be in the shipped tarball at all, since it is a generated file. Additionally, `make dist` is used to make RPMs, so they are also broken. zfs-functions was being added to `EXTRA_DIST` in `contrib/initramfs/Makefile.am`. Removing them fixes the problem.

The other two commits are fixes for minor issues that I found along the way.

### How Has This Been Tested?
Run this:
```
./autogen.sh
./configure
make dist
tar tvf zfs*.tar.gz | grep zfs-functions
```

Expected results:
zfs-functions.in exists, and zfs-functions does not.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by tonyhutter** [comment]
@rlaager FYI - I'm actually testing a fix for this right now:
```patch
diff -ru zfs-0.8.3-orig/contrib/initramfs/Makefile.am zfs-0.8.3/contrib/initramfs/Makefile.am
--- zfs-0.8.3-orig/contrib/initramfs/Makefile.am        2020-01-23 08:50:04.625999986 -0800
+++ zfs-0.8.3/contrib/initramfs/Makefile.am     2020-02-04 17:12:21.074991795 -0800
@@ -6,15 +6,12 @@
 SUBDIRS = hooks scripts

 EXTRA_DIST = \
-       $(top_srcdir)/etc/init.d/zfs \
-       $(top_srcdir)/etc/init.d/zfs-functions \
+       $(top_srcdir)/etc/init.d/zfs.in \
+       $(top_srcdir)/etc/init.d/zfs-functions.in \
        $(top_srcdir)/contrib/initramfs/conf.d/zfs \
        $(top_srcdir)/contrib/initramfs/conf-hooks.d/zfs \
        $(top_srcdir)/contrib/initramfs/README.initramfs.markdown

-$(top_srcdir)/etc/init.d/zfs $(top_srcdir)/etc/init.d/zfs-functions:
-       $(MAKE) -C $(top_srcdir)/etc/init.d zfs zfs-functions
-
 install-initrdSCRIPTS: $(EXTRA_DIST)
        for d in conf.d conf-hooks.d scripts/local-top; do \
                $(MKDIR_P) $(DESTDIR)$(initrddir)/$$d; \
```
I've already built a new `zfs-0.8.3-2.el6.x86_64.rpm` and a new source RPM (containing the patch) and am testing out the new repo to make sure installs/upgrades work ok.  If all goes well, I plan to open a PR and push out the RPMs ASAP.

**Comment by tonyhutter** [comment]
Overall I'm fine with what you have here, but let me sanity test it on Centos 6 just to be sure.  I'll report back today with results.

File: etc/Makefile.am [hunk 1]
```diff
@@ -1,2 +1,2 @@
-SUBDIRS = zfs sudoers.d $(ZFS_INIT_SYSTEMD) $(ZFS_INIT_SYSV) $(ZFS_MODULE_LOAD)
+SUBDIRS = init.d zfs sudoers.d $(ZFS_INIT_SYSTEMD) $(ZFS_INIT_SYSV) $(ZFS_MODULE_LOAD)
```
Comment (behlendorf) [inline etc/Makefile.am]: I believe `init.d` _should_ already be included here by `$(ZFS_INIT_SYSV)` as long as `--enable-sysvinit` was specified.  Everything should be included by `make dist` since they're both listed in the `DIST_SUBDIRS` target.

File: etc/init.d/Makefile.am [hunk 2]
```diff
@@ -15,7 +15,7 @@ EXTRA_DIST = \
 	$(top_srcdir)/etc/init.d/zfs-zed.in \
 	$(top_srcdir)/etc/init.d/zfs.in
 
-$(init_SCRIPTS) $(initconf_SCRIPTS) $(initcommon_SCRIPTS):%:%.in
+$(init_SCRIPTS) $(initconf_SCRIPTS) $(initcommon_SCRIPTS):%:%.in Makefile
```
Comment (behlendorf) [inline etc/init.d/Makefile.am]: I'm not sure I understand why this change is needed.

Comment (rlaager) [reply etc/init.d/Makefile.am]: As described in the commit message:
https://github.com/zfsonlinux/zfs/pull/9955/commits/77acb93b16f3ced9adfc77184275d395c4e7e26a

> This brings it in line with the example:
> https://www.gnu.org/software/automake/manual/html_node/Scripts.html
>
> This way, if the substitution code is changed, they should update.

It's just a correctness thing I noticed while debugging the actual issue. If you don't want it, it's not that big of a deal and I can strip that commit from the stack.

Comment (behlendorf) [reply etc/init.d/Makefile.am]: Ahh, I see my mistake.  I should have looked at the commit message for each change and not the squashed PR.  That makes sense and I think we should keep it.

File: etc/Makefile.am (hunk 1 above)
Comment (rlaager) [reply etc/Makefile.am]: @tonyhutter This was your change. Any thoughts? I agree with Brian here. See config/user-sysvinit.m4. I think this can be removed. I'm pushing a new version of the PR with it removed.

Comment (tonyhutter) [reply etc/Makefile.am]: Correct, you've structured your patch in a way where it's not needed.

Comment (tonyhutter) [reply etc/Makefile.am]: Hmm... maybe not.  I'm seeing no build failures when I manually build your patches on top of zfs-0.8-release on Centos 6, but the build failures on buildbot look like what I was seeing before I added in the `SUBDIRS = init.d` line.  Try adding it back in and see if the build failures go away.

Comment (behlendorf) [reply etc/Makefile.am]: It looks like the fix here is going to need to be a little more involved.  We're going to need to only build the `contrib/initramfs` directory when `--enable_sysvinit` is set, and hence we know the `zfs-functions` files were built.  Adding, the `SUBDIRS = init.d` effectively works around this by always building them which isn't ideal.

Comment (rlaager) [reply etc/Makefile.am]: That doesn’t sound right either. initramfs doesn’t require sysvinit. We will have to find some solution. I’ll take a look. zfs-functions may need to be broken out into a separate directory or something.

Comment (rlaager) [reply etc/Makefile.am]: So there is already an `etc/zfs` directory. It's not obvious why `zfs-functions` isn't in there. I suspect it is historical. I moved it. How does this version work?

I also found that there is duplicated code in the initramfs from before `zfs-functions` was substituted, so I removed that.

**Comment by tonyhutter** [comment]
Looks like there's some build errors from the bots:
```
make[6]: Entering directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/hooks'
/usr/bin/sed -e 's,@sbindir\@,/sbin,g' \
	-e 's,@sysconfdir\@,/etc,g' \
	-e 's,@udevdir\@,/lib/udev,g' \
	-e 's,@udevruledir\@,/lib/udev/rules.d,g' \
	-e 's,@mounthelperdir\@,/sbin,g' \
	zfs.in >'zfs'
make[6]: Leaving directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/hooks'
Making all in scripts
make[6]: Entering directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/scripts'
Making all in local-top
make[7]: Entering directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/scripts/local-top'
make[7]: Nothing to be done for 'all'.
make[7]: Leaving directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/scripts/local-top'
make[7]: Entering directory '/tmp/zfs-build-buildbot-gLUV4QTX/BUILD/zfs-0.8.0/contrib/initramfs/scripts'
make[7]: *** No rule to make target 'zfs', needed by 'all-am'.  Stop.
```

**Comment by rlaager** [comment]
I think this is working now. Thoughts on which commits to squash together (if at all)? For example, should I squash Tony's "Fix zfs-functions packaging bug" and my two "Move" commits together?

**Comment by tonyhutter** [comment]
@rlaager thanks, I'll give this a try on Centos 6 on top of zfs-0.8-release.  I'd recommend squashing this into one commit and removing my "signed-off-by Tony Hutter" line since the majority of this code is yours.

**Comment by rlaager** [comment]
I tend to be in the  "use smaller, cleanly separated commits" camp. There's a fair amount of this that has nothing to do with fixing the bug. If you really want them all squashed, we can. Otherwise, I've squashed some and reordered.
